Add InplaceOrView boxed kernel

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64654
* __->__ #63878

See https://github.com/pytorch/pytorch/issues/64407, https://github.com/pytorch/pytorch/issues/62032 for context:

In this PR:
 - Add boxed kernel by replicating `gen_inplace_or_view`'s logic that is ONLY for use with the Autograd not-implemented kernel
   - Unlike `gen_inplace_or_view` we always pass a view_func to as_view in order to ensure that an "derivative is not implemented" error is raised even if an in-place update is performed on the view. Without the `view_func`, the CopySlice + AsStridedBackward nodes would replace the NotImplemented node.
   - This limitation makes it impossible to use this node for general use
   - view relationship must be between first input (must be tensor) and first output (may be tensor or vec of tensor)
   - do not support non-differentiable views (_values, _indices, view.dtype) - view relationship is always fw and bw differentiable
 - ~~Adds the macro `#define REGISTER_AUTOGRAD_NOT_IMPLEMENTED_FALLBACK(ns, op)` to be the interface for this feature:~~
   - ~~static initialization can be slowed down(? not measured) if there are many registrations, because each line translates to 2 library calls but the workaround is just to manually use the two functions `AutogradNotImplementedFallback` and `ADInplaceOrViewFallback` and call `m.impl`.~~
 - Adds testing:
    - for views: view relationship created
      -  performing in-place operation on the view, raises properly
      - trying to create two view relationships is not allowed,
      - single view relationship but not first input/first output should error
      - view relation created properly for tensor vector output
    - for in-place:
      - version count bump
      - triggers rebase_history
      - multiple mutations is okay and also updates version counter
 - DONE (follow up): Update tutorials for adding  third-party operators (and document the above limitations)
   - [x] https://github.com/pytorch/tutorials/pull/1677
 - TODO (follow up): Look at torch-audio/torch-vision and identify places where this can simplify existing code
 - TODO (follow up): Make it a global fallback. See: https://github.com/pytorch/pytorch/issues/65151

From the previous PR:
 - `#include <torch/csrc/autograd/autograd_not_implemented_fallback.h>` has been added to `torch/csrc/api/include/torch/autograd.h`

   
EDIT: Made it more clear what is introduced in this PR and moved some more contextual stuff into the issue itself

Differential Revision: [D30901714](https://our.internmc.facebook.com/intern/diff/D30901714)

soulitzer committed:
(WIP) Add InplaceOrView boxed fallback

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Add InplaceOrView boxed fallback"


Benchmarking coming soon

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Add InplaceOrView boxed fallback"


Benchmarking coming soon

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Add InplaceOrView boxed fallback"


Benchmarking coming soon

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Add InplaceOrView boxed fallback"


Benchmarking coming soon

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Add InplaceOrView boxed fallback"


Benchmarking coming soon

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Add InplaceOrView boxed fallback"


Benchmarking coming soon

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Add InplaceOrView boxed fallback"


Benchmarking coming soon

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Add InplaceOrView boxed fallback"


Benchmarking coming soon

[ghstack-poisoned]

soulitzer committed:
Update on "Add InplaceOrView boxed fallback"


Fixes https://github.com/pytorch/pytorch/issues/64407
Benchmarking ADInplaceOrView kernel: https://github.com/pytorch/pytorch/pull/64021

[ghstack-poisoned]

soulitzer committed:
Update on "Add InplaceOrView boxed fallback"


Fixes https://github.com/pytorch/pytorch/issues/64407
Benchmarking ADInplaceOrView kernel: https://github.com/pytorch/pytorch/pull/64021

[ghstack-poisoned]

soulitzer committed:
Update on "Add InplaceOrView boxed fallback"


Fixes https://github.com/pytorch/pytorch/issues/64407
Benchmarking ADInplaceOrView kernel: https://github.com/pytorch/pytorch/pull/64021

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -186,4 +186,121 @@ torch::CppFunction autogradNotImplementedFallback() {
+  // We can assume the view is always the first return, and that there will only be a single return
+  // TODO: Document the limitations of this kernel for custom op writers
+  TORCH_INTERNAL_ASSERT((aliased_input_idx == -1 && aliased_output_idx == -1) ||
+    (aliased_input_idx == 0 && aliased_output_idx == 0))
```
An assert here probably isn't the best way to do it; rather it should be a TORCH_CHECK with a longer message saying what to do if you hit it

ezyang commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -186,4 +186,121 @@ torch::CppFunction autogradNotImplementedFallback() {
+  for (const auto i : c10::irange(num_returns)) {
+    const auto& alias_info = returns[i].alias_info();
+    if (alias_info.has_value() && !alias_info->isWrite()) {
+      AT_ASSERT(
```
Nit: TORCH_CHECK

ezyang commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -186,4 +186,121 @@ torch::CppFunction autogradNotImplementedFallback() {
+      // - See NOTE: [ Limitations of ADInplaceOrView boxed kernel ]
+      // - We have to make another copy because the same lambda can be used twice! e.g., double backward
+      std::vector<c10::IValue> stack_args_copy_copy(stack_args_copy);
+      stack_args_copy_copy.at(0) = t;
```
If you're always going to override the 0th index, maybe don't record it in `stack_args_copy_copy` in the first place

ezyang commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -186,4 +186,121 @@ torch::CppFunction autogradNotImplementedFallback() {
+    view_func = [=](const at::Tensor& t) {
+      // - Maybe add a test for this...
+      // - See NOTE: [ Limitations of ADInplaceOrView boxed kernel ]
+      // - We have to make another copy because the same lambda can be used twice! e.g., double backward
```
not just used twice, the lambda can be used concurrently from other threads

ezyang commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -186,4 +186,121 @@ torch::CppFunction autogradNotImplementedFallback() {
+  }
+
+  if (need_view_func) {
+    view_func = [=](const at::Tensor& t) {
```
I'd prefer it if you explicitly said what you were value capturing here

ezyang commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -186,4 +186,121 @@ torch::CppFunction autogradNotImplementedFallback() {
   return torch::CppFunction::makeFromBoxedFunction<&autogradNotImplementedFallbackImpl>();
 }
 
+void ADInplaceOrViewFallbackImpl(const c10::OperatorHandle& op, c10::DispatchKeySet dispatch_keys, torch::jit::Stack* stack) {
```
it would be good to cross-ref this with the relevant codegen file, which this has to be kept in synchronization with

ezyang commented:
One high level remark: the boxed fallback here has a lot more utility for *inplace* functions (which I imagine people are willing to write custom functions for) as opposed to *view* functions (which are a fairly closed ecosystem and most people will just use `as_strided` if they need something weird)

ezyang commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -186,4 +186,121 @@ torch::CppFunction autogradNotImplementedFallback() {
+        const c10::IValue& aliased_input_iv = (*stack)[stack_start + i]; // get a reference to an ivalue on the stack
+        TORCH_CHECK(aliased_input_iv.isTensor());
+        // copy assignment
+        aliased_input = aliased_input_iv.toTensor();
```
not sure you actually need this refcount bump

ezyang commented:
Tests???

soulitzer commented:
One point in favor of keeping the logic for handling views could be that if we only support in-place, maybe we still want the same logic detecting if there are non-write aliases in inputs/outputs in order to raise an error/warning

soulitzer committed:
Update on "Add InplaceOrView boxed kernel"


See https://github.com/pytorch/pytorch/issues/64407, https://github.com/pytorch/pytorch/issues/62032 for some more context
 - Still prefer boxed over templated
   - unless boxed is insanely slow we always prefer boxed because it's only a default anyway
 - It does not seem like a clear win to replace (see discussion at bottom)

In this PR:
 - Add boxed kernel by replicating gen_inplace_or_view logic
 - Introduces the macro `#define REGISTER_AUTOGRAD_NOT_IMPLEMENTED_FALLBACK(ns, op)` (name tbd) to be the interface for this feature:
   - static initialization can be slowed down(? not measured) if there are many registrations, because each line translates to 2 library calls but the workaround is just to manually use the two functions `AutogradNotImplementedFallback` and `ADInplaceOrViewFallback` and call `m.impl`.
 - we support limited number of cases:
   - Just like gen_inplace_or_view logic 
      - the base/view tensors must be first tensor input or first tensor/tensorlist output
      - can only have one view relationship
     - op cannot be both inplace and view (we don't have true inplace views)
   - More limitations:
     - do not support non-differentiable views (_values, _indices, view.dtype)
     - do not support ops that always need metadata change (view_as_complex, view_as_real, _conj, _neg_view)
   - (...)
 - TODO (follow up): understand/document the above limitations better
 - TODO (this PR): update tests
   
Testing on CI by replacing codegen here: https://github.com/pytorch/pytorch/pull/64621
 - makes some workarounds to get ci to pass for everything
   - non-differentiable views: _values, _indices, view_dtype
   - https://github.com/pytorch/pytorch/issues/64059
     - the workaround should not be landed
     - this should probably be fixed in codegen 
     
Benchmarking here: https://github.com/pytorch/pytorch/pull/64021
 - compares templated, boxed, codegen
 - things were mostly as expected
 - there is also more benchmarking in the issues linked above
 - more ops could potentially be tested but what we measure here is the worst case (the op itself does little work, and the overhead is most visible).
 
Is templated good enough?
 - Probably not:
   - the gap between boxed and codegen isn't THAT big anyway
   - boxed is good enough as a default option, users can always write their own ops
   - templated is indeed faster because...
     - we can precompute certain values
     - other overhead of boxing?
   - even if it is fast, there is not too much benefit to replacing the codegen
     - the templated code is not ideal to work with (maybe there is room for improvement?)

Should we replace? (boxed)
 - pros of replacing
   - reduced logic duplication between codegen and this kernel
   - reduced binary size, faster compilation (not measured, but expected to be small based on previous measurements for VariableType)
 - cons of replacing
   - 10-15% slowdown for small view/inplace ops (boxed)
     - this could be potentially bad if a model uses a lot of these
     - haven't tested for larger tensor sizes though
   - there is some more work to do
     - other edge cases like non-differentiable ops (see testing section above)


[ghstack-poisoned]

soulitzer commented:
In this update, I'm making it so that we always pass a `view_func` to `as_view`. This is needed because when in-place is performed on a view, if `view_func` was not specified on the creation of that view, `AsStridedBackward` and `CopySlices` looks at the strides of the output node, and we no longer raise the "derivative not implemented" error that we desired. This also means that this kernel is to be specifically used with autograd not implemeneted kernel only, but that is okay because it we already decided that we won't be replacing the codegened kernels.

Some basic testing has also been added:
1) for views: view relationship created
  a) on in-place error is raised
  b) double view is not allowed, 
  c) TODO: single view relationship but not first input/first output
2) for in-place: version count bump, triggers rebase_history
 a two in-place is okay

soulitzer committed:
Update on "Add InplaceOrView boxed kernel"


See https://github.com/pytorch/pytorch/issues/64407, https://github.com/pytorch/pytorch/issues/62032 for context:

In this PR:
 - Add boxed kernel by replicating `gen_inplace_or_view`'s logic that is ONLY for use with the Autograd not-implemented kernel
   - Unlike `gen_inplace_or_view` we always pass a view_func to as_view in order to ensure that an "derivative is not implemented" error is raised even if an in-place update is performed on the view. Without the `view_func`, the CopySlice + AsStridedBackward nodes would replace the NotImplemented node.
   - This limitation makes it impossible to use this node for general use
   - view relationship must be between first input (must be tensor) and first output (may be tensor or vec of tensor)
   - do not support non-differentiable views (_values, _indices, view.dtype) - view relationship is always fw and bw differentiable
 - Adds the macro `#define REGISTER_AUTOGRAD_NOT_IMPLEMENTED_FALLBACK(ns, op)` to be the interface for this feature:
   - static initialization can be slowed down(? not measured) if there are many registrations, because each line translates to 2 library calls but the workaround is just to manually use the two functions `AutogradNotImplementedFallback` and `ADInplaceOrViewFallback` and call `m.impl`.
 - Adds testing:
    - for views: view relationship created
      -  performing in-place operation on the view, raises properly
      - trying to create two view relationships is not allowed,
      - TODO (this PR): single view relationship but not first input/first output should error
      - TODO (this PR): view relation created properly for tensor vector output
    - for in-place:
      - version count bump
      - triggers rebase_history
      - multiple mutations is okay and also updates version counter
 - TODO (follow up): Update tutorials for adding  third-party operators (and document the above limitations)
 - TODO (follow up): Look at torch-audio/torch-vision and identify places where this can simplify existing code
   
EDIT: Made it more clear what is introduced in this PR and moved some more contextual stuff into the issue itself

[ghstack-poisoned]

albanD commented on test/cpp/api/autograd.cpp:
```diff
@@ -877,7 +877,7 @@ TEST(CustomAutogradTest, BackwardWithCreateGraphWarns) {
  * - check_inplace logic
  * - view ops
- * - TODO: Tests for NDEBUG checks?
+ * - TODO: Tests for NDEBUG checks? Don't need for now because CI doesn't test NDEBUG builds.
```
nit: doesn't test without NDEBUG :D 
double negations!

soulitzer replied:
Whoops!

soulitzer committed:
Update on "Add InplaceOrView boxed kernel"


See https://github.com/pytorch/pytorch/issues/64407, https://github.com/pytorch/pytorch/issues/62032 for context:

In this PR:
 - Add boxed kernel by replicating `gen_inplace_or_view`'s logic that is ONLY for use with the Autograd not-implemented kernel
   - Unlike `gen_inplace_or_view` we always pass a view_func to as_view in order to ensure that an "derivative is not implemented" error is raised even if an in-place update is performed on the view. Without the `view_func`, the CopySlice + AsStridedBackward nodes would replace the NotImplemented node.
   - This limitation makes it impossible to use this node for general use
   - view relationship must be between first input (must be tensor) and first output (may be tensor or vec of tensor)
   - do not support non-differentiable views (_values, _indices, view.dtype) - view relationship is always fw and bw differentiable
 - Adds the macro `#define REGISTER_AUTOGRAD_NOT_IMPLEMENTED_FALLBACK(ns, op)` to be the interface for this feature:
   - static initialization can be slowed down(? not measured) if there are many registrations, because each line translates to 2 library calls but the workaround is just to manually use the two functions `AutogradNotImplementedFallback` and `ADInplaceOrViewFallback` and call `m.impl`.
 - Adds testing:
    - for views: view relationship created
      -  performing in-place operation on the view, raises properly
      - trying to create two view relationships is not allowed,
      - single view relationship but not first input/first output should error
      - view relation created properly for tensor vector output
    - for in-place:
      - version count bump
      - triggers rebase_history
      - multiple mutations is okay and also updates version counter
 - TODO (follow up): Update tutorials for adding  third-party operators (and document the above limitations)
 - TODO (follow up): Look at torch-audio/torch-vision and identify places where this can simplify existing code
   
EDIT: Made it more clear what is introduced in this PR and moved some more contextual stuff into the issue itself

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30901714).

ezyang commented:
@albanD I'm going to let you take a pop at this review first

soulitzer committed:
Update on "Add InplaceOrView boxed kernel"


See https://github.com/pytorch/pytorch/issues/64407, https://github.com/pytorch/pytorch/issues/62032 for context:

In this PR:
 - Add boxed kernel by replicating `gen_inplace_or_view`'s logic that is ONLY for use with the Autograd not-implemented kernel
   - Unlike `gen_inplace_or_view` we always pass a view_func to as_view in order to ensure that an "derivative is not implemented" error is raised even if an in-place update is performed on the view. Without the `view_func`, the CopySlice + AsStridedBackward nodes would replace the NotImplemented node.
   - This limitation makes it impossible to use this node for general use
   - view relationship must be between first input (must be tensor) and first output (may be tensor or vec of tensor)
   - do not support non-differentiable views (_values, _indices, view.dtype) - view relationship is always fw and bw differentiable
 - Adds the macro `#define REGISTER_AUTOGRAD_NOT_IMPLEMENTED_FALLBACK(ns, op)` to be the interface for this feature:
   - static initialization can be slowed down(? not measured) if there are many registrations, because each line translates to 2 library calls but the workaround is just to manually use the two functions `AutogradNotImplementedFallback` and `ADInplaceOrViewFallback` and call `m.impl`.
 - Adds testing:
    - for views: view relationship created
      -  performing in-place operation on the view, raises properly
      - trying to create two view relationships is not allowed,
      - single view relationship but not first input/first output should error
      - view relation created properly for tensor vector output
    - for in-place:
      - version count bump
      - triggers rebase_history
      - multiple mutations is okay and also updates version counter
 - TODO (follow up): Update tutorials for adding  third-party operators (and document the above limitations)
   - https://github.com/pytorch/tutorials/pull/1677
 - TODO (follow up): Look at torch-audio/torch-vision and identify places where this can simplify existing code

From the previous PR:
 - `#include <torch/csrc/autograd/autograd_not_implemented_fallback.h>` has been added to `torch/csrc/api/include/torch/autograd.h`

   
EDIT: Made it more clear what is introduced in this PR and moved some more contextual stuff into the issue itself

Differential Revision: [D30901714](https://our.internmc.facebook.com/intern/diff/D30901714)

[ghstack-poisoned]

soulitzer committed:
Update on "Add InplaceOrView boxed kernel"


See https://github.com/pytorch/pytorch/issues/64407, https://github.com/pytorch/pytorch/issues/62032 for context:

In this PR:
 - Add boxed kernel by replicating `gen_inplace_or_view`'s logic that is ONLY for use with the Autograd not-implemented kernel
   - Unlike `gen_inplace_or_view` we always pass a view_func to as_view in order to ensure that an "derivative is not implemented" error is raised even if an in-place update is performed on the view. Without the `view_func`, the CopySlice + AsStridedBackward nodes would replace the NotImplemented node.
   - This limitation makes it impossible to use this node for general use
   - view relationship must be between first input (must be tensor) and first output (may be tensor or vec of tensor)
   - do not support non-differentiable views (_values, _indices, view.dtype) - view relationship is always fw and bw differentiable
 - Adds the macro `#define REGISTER_AUTOGRAD_NOT_IMPLEMENTED_FALLBACK(ns, op)` to be the interface for this feature:
   - static initialization can be slowed down(? not measured) if there are many registrations, because each line translates to 2 library calls but the workaround is just to manually use the two functions `AutogradNotImplementedFallback` and `ADInplaceOrViewFallback` and call `m.impl`.
 - Adds testing:
    - for views: view relationship created
      -  performing in-place operation on the view, raises properly
      - trying to create two view relationships is not allowed,
      - single view relationship but not first input/first output should error
      - view relation created properly for tensor vector output
    - for in-place:
      - version count bump
      - triggers rebase_history
      - multiple mutations is okay and also updates version counter
 - TODO (follow up): Update tutorials for adding  third-party operators (and document the above limitations)
   - https://github.com/pytorch/tutorials/pull/1677
 - TODO (follow up): Look at torch-audio/torch-vision and identify places where this can simplify existing code

From the previous PR:
 - `#include <torch/csrc/autograd/autograd_not_implemented_fallback.h>` has been added to `torch/csrc/api/include/torch/autograd.h`

   
EDIT: Made it more clear what is introduced in this PR and moved some more contextual stuff into the issue itself

Differential Revision: [D30901714](https://our.internmc.facebook.com/intern/diff/D30901714)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -186,4 +185,134 @@ torch::CppFunction autogradNotImplementedFallback() {
+  at::Tensor aliased_input;
+
+  int64_t aliased_output_idx = -1;
+  for (const auto i : c10::irange(num_returns)) {
```
A simple way to speed this up would be to actually do:
```cpp
#ifdef NDEBUG
  int64_t aliased_output_idx = returns[i].alias_info().has_value() && !returns[i].alias_info()->isWrite() ? 0 : -1;
#else
  // The current logic including the TORCH_CHECK below to check that you get a 0
#endif

```

soulitzer replied:
This makes it so that if user has alias in non-first input those are silently ignored unless they are in debug mode. Could that be problematic? This looks like a good optimization but I don't know if this will speed up the code by tooo much, so I'm okay with having slower code here but with slightly better UX since that's the tradeoff we are already making in general with this boxed kernel.

albanD replied:
Ho yes we don't need to do that. But when we were discussing optimizations for this code, this actually removes the need for traversal of all the inputs.

ezyang replied:
I don't think it's appropriate to disable this check in DEBUG mode, as it is general out-of-tree user code that we are validating over.

ezyang replied:
A better way to apply this optimization is to precompute more information in the schema, so that, e.g., we could query a single field on the schema to find out more information that we need. Something to think about as we start exposing function schema as a user inspectable data structure in primtorch introspection track.

albanD commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -186,4 +185,134 @@ torch::CppFunction autogradNotImplementedFallback() {
+        const c10::IValue& aliased_input_iv = (*stack)[stack_start + i]; // get a reference to an ivalue on the stack
+        TORCH_CHECK(aliased_input_iv.isTensor());
+        // copy assignment (TODO: how to avoid this extra reference count bump??)
+        // It doesn't seem like we can save merely a pointer to Tensor because it would
```
Why would it get dropped from the stack? During the redispatch below?
I think this extra call is ok tbh. Maybe clean up the comment a bit?

soulitzer replied:
Yeah, I think it should get dropped from the stack during redispatch. Will just clean up the comment for now.

soulitzer commented on test/cpp/api/autograd.cpp:
```diff
@@ -896,9 +896,12 @@ std::tuple<torch::Tensor, torch::Tensor> two_arg_inplace_op(const torch::Tensor&
-  auto self_view = self.view(-1);
-  auto other_view = other.view(-1);
-  return std::tuple<torch::Tensor, torch::Tensor>(self_view, other_view);
+  return std::tuple<torch::Tensor, torch::Tensor>(self, other);
```
I think we are ok because alias info still says that the output aliases the input. We are basically just doing `alias`.
If the function itself returns `.view`, then it'd still have a view relationship even if the op is registered in a composite way, and we wouldn't be testing it properly.

soulitzer commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -1,3 +1,7 @@
+# Generates ADInplaceOrViewType.h/cpp
+# **If any changes are being made to the ADInplaceOrView codegen please also check
```
Added.
 ** was for grabbing attention, changed it do an uppercase NOTE

soulitzer commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -186,4 +185,134 @@ torch::CppFunction autogradNotImplementedFallback() {
+  // - (TODO?) enforce that the same op cannot be both a view and
+  //   that is not allowed in the gen_inplace_or_view logic
+  //
+  // TODO: We could just link to the documentation once we have that
```
Oh right, adding the link

soulitzer commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -44,7 +44,8 @@ void _foreach_tensor(
 void autogradNotImplementedFallbackImpl(const c10::OperatorHandle& op, c10::DispatchKeySet dispatch_keys, torch::jit::Stack* stack) {
-  // Mimics the logic of a VariableType NotImplemented kernel
+  // Mimics a subset of the logic of a VariableType NotImplemented kernel
+  // - see gen_variable_type.py
```
bullet point, if it can be confused to have another meaning though I can remove it

albanD replied:
A bullet point with only 1 thing is a bit weird I think.

soulitzer replied:
Hehe I guess so, I fixed the cases below with one bullet point by combining them. This one has 3 now, so I will keep it.

albanD replied:
I see just 1 right now. Is there an update of the code that is missing from the PR?

soulitzer replied:
Oh whoops, I was looking at the InplaceOrView one

soulitzer committed:
Update on "Add InplaceOrView boxed kernel"


See https://github.com/pytorch/pytorch/issues/64407, https://github.com/pytorch/pytorch/issues/62032 for context:

In this PR:
 - Add boxed kernel by replicating `gen_inplace_or_view`'s logic that is ONLY for use with the Autograd not-implemented kernel
   - Unlike `gen_inplace_or_view` we always pass a view_func to as_view in order to ensure that an "derivative is not implemented" error is raised even if an in-place update is performed on the view. Without the `view_func`, the CopySlice + AsStridedBackward nodes would replace the NotImplemented node.
   - This limitation makes it impossible to use this node for general use
   - view relationship must be between first input (must be tensor) and first output (may be tensor or vec of tensor)
   - do not support non-differentiable views (_values, _indices, view.dtype) - view relationship is always fw and bw differentiable
 - ~~Adds the macro `#define REGISTER_AUTOGRAD_NOT_IMPLEMENTED_FALLBACK(ns, op)` to be the interface for this feature:~~
   - ~~static initialization can be slowed down(? not measured) if there are many registrations, because each line translates to 2 library calls but the workaround is just to manually use the two functions `AutogradNotImplementedFallback` and `ADInplaceOrViewFallback` and call `m.impl`.~~
 - Adds testing:
    - for views: view relationship created
      -  performing in-place operation on the view, raises properly
      - trying to create two view relationships is not allowed,
      - single view relationship but not first input/first output should error
      - view relation created properly for tensor vector output
    - for in-place:
      - version count bump
      - triggers rebase_history
      - multiple mutations is okay and also updates version counter
 - DONE (follow up): Update tutorials for adding  third-party operators (and document the above limitations)
   - [x] https://github.com/pytorch/tutorials/pull/1677
 - TODO (follow up): Look at torch-audio/torch-vision and identify places where this can simplify existing code
 - TODO (follow up): Make it a global fallback. See: https://github.com/pytorch/pytorch/issues/65151

From the previous PR:
 - `#include <torch/csrc/autograd/autograd_not_implemented_fallback.h>` has been added to `torch/csrc/api/include/torch/autograd.h`

   
EDIT: Made it more clear what is introduced in this PR and moved some more contextual stuff into the issue itself

Differential Revision: [D30901714](https://our.internmc.facebook.com/intern/diff/D30901714)

[ghstack-poisoned]

soulitzer committed:
Update on "Add InplaceOrView boxed kernel"


See https://github.com/pytorch/pytorch/issues/64407, https://github.com/pytorch/pytorch/issues/62032 for context:

In this PR:
 - Add boxed kernel by replicating `gen_inplace_or_view`'s logic that is ONLY for use with the Autograd not-implemented kernel
   - Unlike `gen_inplace_or_view` we always pass a view_func to as_view in order to ensure that an "derivative is not implemented" error is raised even if an in-place update is performed on the view. Without the `view_func`, the CopySlice + AsStridedBackward nodes would replace the NotImplemented node.
   - This limitation makes it impossible to use this node for general use
   - view relationship must be between first input (must be tensor) and first output (may be tensor or vec of tensor)
   - do not support non-differentiable views (_values, _indices, view.dtype) - view relationship is always fw and bw differentiable
 - ~~Adds the macro `#define REGISTER_AUTOGRAD_NOT_IMPLEMENTED_FALLBACK(ns, op)` to be the interface for this feature:~~
   - ~~static initialization can be slowed down(? not measured) if there are many registrations, because each line translates to 2 library calls but the workaround is just to manually use the two functions `AutogradNotImplementedFallback` and `ADInplaceOrViewFallback` and call `m.impl`.~~
 - Adds testing:
    - for views: view relationship created
      -  performing in-place operation on the view, raises properly
      - trying to create two view relationships is not allowed,
      - single view relationship but not first input/first output should error
      - view relation created properly for tensor vector output
    - for in-place:
      - version count bump
      - triggers rebase_history
      - multiple mutations is okay and also updates version counter
 - DONE (follow up): Update tutorials for adding  third-party operators (and document the above limitations)
   - [x] https://github.com/pytorch/tutorials/pull/1677
 - TODO (follow up): Look at torch-audio/torch-vision and identify places where this can simplify existing code
 - TODO (follow up): Make it a global fallback. See: https://github.com/pytorch/pytorch/issues/65151

From the previous PR:
 - `#include <torch/csrc/autograd/autograd_not_implemented_fallback.h>` has been added to `torch/csrc/api/include/torch/autograd.h`

   
EDIT: Made it more clear what is introduced in this PR and moved some more contextual stuff into the issue itself

Differential Revision: [D30901714](https://our.internmc.facebook.com/intern/diff/D30901714)

[ghstack-poisoned]

albanD reviewed (approved):
LGTM

soulitzer commented:
> @albanD I'm going to let you take a pop at this review first
 
@ezyang did you want to take another look? Would appreciate any comments!

soulitzer committed:
Update on "Add InplaceOrView boxed kernel"


See https://github.com/pytorch/pytorch/issues/64407, https://github.com/pytorch/pytorch/issues/62032 for context:

In this PR:
 - Add boxed kernel by replicating `gen_inplace_or_view`'s logic that is ONLY for use with the Autograd not-implemented kernel
   - Unlike `gen_inplace_or_view` we always pass a view_func to as_view in order to ensure that an "derivative is not implemented" error is raised even if an in-place update is performed on the view. Without the `view_func`, the CopySlice + AsStridedBackward nodes would replace the NotImplemented node.
   - This limitation makes it impossible to use this node for general use
   - view relationship must be between first input (must be tensor) and first output (may be tensor or vec of tensor)
   - do not support non-differentiable views (_values, _indices, view.dtype) - view relationship is always fw and bw differentiable
 - ~~Adds the macro `#define REGISTER_AUTOGRAD_NOT_IMPLEMENTED_FALLBACK(ns, op)` to be the interface for this feature:~~
   - ~~static initialization can be slowed down(? not measured) if there are many registrations, because each line translates to 2 library calls but the workaround is just to manually use the two functions `AutogradNotImplementedFallback` and `ADInplaceOrViewFallback` and call `m.impl`.~~
 - Adds testing:
    - for views: view relationship created
      -  performing in-place operation on the view, raises properly
      - trying to create two view relationships is not allowed,
      - single view relationship but not first input/first output should error
      - view relation created properly for tensor vector output
    - for in-place:
      - version count bump
      - triggers rebase_history
      - multiple mutations is okay and also updates version counter
 - DONE (follow up): Update tutorials for adding  third-party operators (and document the above limitations)
   - [x] https://github.com/pytorch/tutorials/pull/1677
 - TODO (follow up): Look at torch-audio/torch-vision and identify places where this can simplify existing code
 - TODO (follow up): Make it a global fallback. See: https://github.com/pytorch/pytorch/issues/65151

From the previous PR:
 - `#include <torch/csrc/autograd/autograd_not_implemented_fallback.h>` has been added to `torch/csrc/api/include/torch/autograd.h`

   
EDIT: Made it more clear what is introduced in this PR and moved some more contextual stuff into the issue itself

Differential Revision: [D30901714](https://our.internmc.facebook.com/intern/diff/D30901714)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30901714).

soulitzer committed:
Update on "Add InplaceOrView boxed kernel"


See https://github.com/pytorch/pytorch/issues/64407, https://github.com/pytorch/pytorch/issues/62032 for context:

In this PR:
 - Add boxed kernel by replicating `gen_inplace_or_view`'s logic that is ONLY for use with the Autograd not-implemented kernel
   - Unlike `gen_inplace_or_view` we always pass a view_func to as_view in order to ensure that an "derivative is not implemented" error is raised even if an in-place update is performed on the view. Without the `view_func`, the CopySlice + AsStridedBackward nodes would replace the NotImplemented node.
   - This limitation makes it impossible to use this node for general use
   - view relationship must be between first input (must be tensor) and first output (may be tensor or vec of tensor)
   - do not support non-differentiable views (_values, _indices, view.dtype) - view relationship is always fw and bw differentiable
 - ~~Adds the macro `#define REGISTER_AUTOGRAD_NOT_IMPLEMENTED_FALLBACK(ns, op)` to be the interface for this feature:~~
   - ~~static initialization can be slowed down(? not measured) if there are many registrations, because each line translates to 2 library calls but the workaround is just to manually use the two functions `AutogradNotImplementedFallback` and `ADInplaceOrViewFallback` and call `m.impl`.~~
 - Adds testing:
    - for views: view relationship created
      -  performing in-place operation on the view, raises properly
      - trying to create two view relationships is not allowed,
      - single view relationship but not first input/first output should error
      - view relation created properly for tensor vector output
    - for in-place:
      - version count bump
      - triggers rebase_history
      - multiple mutations is okay and also updates version counter
 - DONE (follow up): Update tutorials for adding  third-party operators (and document the above limitations)
   - [x] https://github.com/pytorch/tutorials/pull/1677
 - TODO (follow up): Look at torch-audio/torch-vision and identify places where this can simplify existing code
 - TODO (follow up): Make it a global fallback. See: https://github.com/pytorch/pytorch/issues/65151

From the previous PR:
 - `#include <torch/csrc/autograd/autograd_not_implemented_fallback.h>` has been added to `torch/csrc/api/include/torch/autograd.h`

   
EDIT: Made it more clear what is introduced in this PR and moved some more contextual stuff into the issue itself

Differential Revision: [D30901714](https://our.internmc.facebook.com/intern/diff/D30901714)

[ghstack-poisoned]

soulitzer committed:
Update on "Add InplaceOrView boxed kernel"


See https://github.com/pytorch/pytorch/issues/64407, https://github.com/pytorch/pytorch/issues/62032 for context:

In this PR:
 - Add boxed kernel by replicating `gen_inplace_or_view`'s logic that is ONLY for use with the Autograd not-implemented kernel
   - Unlike `gen_inplace_or_view` we always pass a view_func to as_view in order to ensure that an "derivative is not implemented" error is raised even if an in-place update is performed on the view. Without the `view_func`, the CopySlice + AsStridedBackward nodes would replace the NotImplemented node.
   - This limitation makes it impossible to use this node for general use
   - view relationship must be between first input (must be tensor) and first output (may be tensor or vec of tensor)
   - do not support non-differentiable views (_values, _indices, view.dtype) - view relationship is always fw and bw differentiable
 - ~~Adds the macro `#define REGISTER_AUTOGRAD_NOT_IMPLEMENTED_FALLBACK(ns, op)` to be the interface for this feature:~~
   - ~~static initialization can be slowed down(? not measured) if there are many registrations, because each line translates to 2 library calls but the workaround is just to manually use the two functions `AutogradNotImplementedFallback` and `ADInplaceOrViewFallback` and call `m.impl`.~~
 - Adds testing:
    - for views: view relationship created
      -  performing in-place operation on the view, raises properly
      - trying to create two view relationships is not allowed,
      - single view relationship but not first input/first output should error
      - view relation created properly for tensor vector output
    - for in-place:
      - version count bump
      - triggers rebase_history
      - multiple mutations is okay and also updates version counter
 - DONE (follow up): Update tutorials for adding  third-party operators (and document the above limitations)
   - [x] https://github.com/pytorch/tutorials/pull/1677
 - TODO (follow up): Look at torch-audio/torch-vision and identify places where this can simplify existing code
 - TODO (follow up): Make it a global fallback. See: https://github.com/pytorch/pytorch/issues/65151

From the previous PR:
 - `#include <torch/csrc/autograd/autograd_not_implemented_fallback.h>` has been added to `torch/csrc/api/include/torch/autograd.h`

   
EDIT: Made it more clear what is introduced in this PR and moved some more contextual stuff into the issue itself

Differential Revision: [D30901714](https://our.internmc.facebook.com/intern/diff/D30901714)

[ghstack-poisoned]

albanD reviewed (approved):
Update to alias analysis looks good!
